src/mbgl/geometry/feature_index.cpp

```diff
@@ -194,4 +210,29 @@ void FeatureIndex::setBucketLayerIDs(const std::string& bucketLeaderID, const st
     bucketLayerIDs[bucketLeaderID] = layerIDs;
 }
 
+void FeatureIndexDynamic::query(std::unordered_map<std::string, std::vector<Feature>>& result,
+                                const mbgl::ScreenLineString& queryGeometry,
+                                const TransformState& state) const {
+    mbgl::WithinBBox queryBox = DefaultBBox;
+    for (const auto& p : queryGeometry) {
+        const LatLng c = screenCoordinateToLatLng(p, state);
+        const Point<double> pm = project(c, state);
+        const Point<int64_t> coord = {int64_t(pm.x), int64_t(pm.y)};
+        mbgl::updateBBox(queryBox, coord);
+    }
+    for (const auto& f : features) {
+        // hit testing
+        mbgl::WithinBBox featureBox = DefaultBBox;
+        for (const auto& p : f.envelope->front()) mbgl::updateBBox(featureBox, p);
+
+        const bool hit = mbgl::boxWithinBox(featureBox, queryBox) || mbgl::boxWithinBox(queryBox, featureBox);
+        if (hit) result[f.feature->sourceLayer].push_back(*f.feature);
```

Comment:
`assert(f.feature)`? is it guaranteed?

Reply:
Indeed other implementations might do mistakes